### ISSUE TITLE
Code implementations are not working

### ISSUE THREAD

**Issue by daehuikim** [rmahari/LePaRD#1]
Hello
I am trying to reproduce the experiments using your code implementation. However, I am currently unable to reproduce any results due to several issues. Could you kindly provide the dependencies or a virtual environment configuration required to run the code?

Additionally, it seems that some of the modules in the code have been deprecated, making it challenging to proceed without further clarification.

Your assistance in resolving these issues would be greatly appreciated.

Thank you for your time and support.

**Comment by dominiksinsaarland** [comment]
@daehuikim Thanks for being in touch, and indeed, this repo was not very polished.

I updated the relevant parts, and tested them locally and it worked for me with a clean environment. Please also have a look at the updated README. If anything still shouldn't work, please let me know!

Cheers,
Dominik

**Comment by daehuikim** [comment]
@dominiksinsaarland Thanks for updating.

Currently, I am conducting experiments using the data provided in the V3 version. However, I noticed that the results I reproduced differ significantly from the values reported in Table 3. For example, when using SBERT-Finetuned on the 10K dataset, my reproduced Recall@1, Recall@10, nDCG, and MAP are 12.66, 46.15, 27.34, and 21.58, respectively, which are quite different from the reported values.

To ensure accuracy, I ran the experiments without modifying the source code you provided. Moreover, the values reported in the updated Table 3 are substantially higher than those achieved when using the "Quote" column (ground truth) directly for retrieval, which I found curious.

If possible, could you provide guidance on how the reported results were obtained or any advice on reproducing them accurately? Any insights you can share would be greatly appreciated.
Thank you for your time and consideration, and I look forward to your response.

Best regards,
Daehee Kim

**Comment by dominiksinsaarland** [comment]
That's odd! I am currently re-running the SBERT-finetuned experiments, will let you know as soon as I have new results. The reported results should be obtained with


num_labels="10000" # change this to 20000 / 50000 for other experiments
python src/model/finetune_sbert.py --n_labels $num_labels
$model_name="finetuned-distilbert-base-uncased-10000"
# run inference
python src/model/run_inference_sbert.py --model_name $model_name n_labels $num_labels # creates folder "predictions-sbert" and saves output there (os.path.basename(model_name) + predictions dev/test)
# evaluate
python src/model/evaluate_run.py --dev_predictions finetuned-distilbert-base-uncased-10000/predictions_devset_10000.json --test_predictions finetuned-distilbert-base-uncased-10000/predictions_testset_10000.json 

will let you know in a few hours if I am able to replicate that

**Comment by dominiksinsaarland** [comment]
Still investigating, currently getting the same results as you have. Will let you know as soon as we have an update here!

**Comment by daehuikim** [comment]
@dominiksinsaarland 
Appreciate for your demonstration! This resource is priceless :)

**Comment by dominiksinsaarland** [comment]
I hope I fixed it -- I assume it was something about how the training data was generated in the "finetune_sbert.py" file. Can you try again? And please start with the relevant files in https://huggingface.co/datasets/rmahari/LePaRD/tree/main (you'll need "top_10000_data.csv.gz" and "passage_dict.json"
Also, I changed the "model.fit(" to "model.old_fit" as we were working with an early version of SBERT when conducting these experiments, that might also have had an influence on the results.

All of this results in the following results:
n labels dev 10000 -- recall @  1 19.78
n labels dev 10000 -- recall @  5 48.97
n labels dev 10000 -- recall @  10 62.75
devset ndcg 39.4
devset map 32.14
95218 95218
n labels dev 10000 -- recall @  1 20.0
n labels dev 10000 -- recall @  5 49.16
n labels dev 10000 -- recall @  10 62.78
testset ndcg 39.54
testset map 32.32

Which seems close enough to what we reported in the paper (gpu / seed etc.). 

Please let me know in case anything doesn't work or your results are still different! Happy to investigate further then.

**Comment by daehuikim** [comment]
@dominiksinsaarland 
Thanks for your investigation.
https://github.com/rmahari/LePaRD/blob/cf9d84534a24c118041061eae7bc3f4f9b69e2a9/src/model/finetune_sbert.py#L21
What is the ```passage_dict``` in this line? This variable is not defined before it used.
Also, Is this about processing data as destination context (sequence of words) to passage_id (number)?

**Comment by daehuikim** [comment]
I checked the changes in the code implementation through the commit history. While the data processing method has changed compared to the previous code, I confirmed that the processed data remains the same as before. Could the performance gap be due to the difference between the ```fit``` and ```old_fit``` method? I’ve also just run a new experiment. If possible, I would greatly appreciate it if you could share the trained model's weights.

**Comment by dominiksinsaarland** [comment]
Sorry about this, fixed the passage_dict file in finetune_sbert.py

Also, I uploaded the trained model here: https://huggingface.co/dominiks/lepard-finetuned-sbert

model_name="dominiks/lepard-finetuned-sbert"
python run_inference_sbert.py --model_name $model_name

and then evaluate

**Comment by daehuikim** [comment]
@dominiksinsaarland Thanks for sharing your checkpoint! I got the same results!

I am still unable to reproduce the fine-tuning results through the code. 
Could there be any differences between the data available in the Dropbox mentioned in the README and the data on Hugging Face?

**Comment by dominiksinsaarland** [comment]
ok, tbh I am a bit lost. I checked and there shouldn't be any difference in the dropbox and huggingface data (otherwise, you also wouldn't be able to replicate the results from the same checkpoint I suppose)

I ran things with the following sentence-transformer version: 
Version: 3.2.0

Could you check if you have the same? Just run "pip show sentence-transformers" in a terminal

Also, I zipped everything in my replication folder (just the 10K labels version). You can download that here. If you have the same sbert version would you want to give this a try?

https://www.dropbox.com/scl/fi/wcsvt1xdzxl4zt6qmeas0/replication_package_lepard_sbert.zip?rlkey=t6krdpo2u7c0w6lj200kk52wj&st=fbdue212&dl=0

python src/model/finetune_sbert.py 

etc.

**Comment by dominiksinsaarland** [comment]
@daehuikim Is it possible that in your case, sentence-bert for some reason doesn't load the modelweights, and then silently initalizes all of them randomly? I ran one test with the distilroberta architecture where I manually set all biases to 0, reset all model weights to be drawn from a normal distribution with mean=0 and std=0.02. The results are: rc@1=10.0; and rc@10=38.88 -- which seems fairly close to the rc@1 of 12.66 and rc@10 of 46.15 you reported earlier.

**Comment by daehuikim** [comment]
@dominiksinsaarland Thanks for your confirmation!

My Sentence-tranformer version was 3.4.1 which is differ from you.
Maybe the initialization can be the matter I will try with your replication folder
Thanks again :)
